Canonicalize runtime asserts

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114509

This allows us to remove quite a few redundant runtime asserts, and potentially a number of guards as well.

On
```
python test/dynamo/test_subclasses.py -k test_unbind
```
we go from
```
inserting runtime assert i0 <= s0
inserting runtime assert 0 <= -i0 + s0
inserting runtime assert i0 + i1 <= s0
inserting runtime assert i0 <= -i1 + s0
inserting runtime assert i0 + i1 + i2 <= s0
inserting runtime assert i0 + i1 <= -i2 + s0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
inserting runtime assert i0 + i1 + i2 + i3 <= s0
inserting runtime assert i0 + i1 + i2 <= -i3 + s0
```
to
```
inserting runtime assert i0 - s0 <= 0
inserting runtime assert i0 + i1 - s0 <= 0
inserting runtime assert i0 + i1 + i2 - s0 <= 0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
```

cc @ezyang

lezcano committed:
Canonicalize the runtime asserts

This allows us to remove quite a few redundant runtime asserts. For
```
python test/dynamo/test_subclasses.py -k test_unbind
```

We go from
```
inserting runtime assert i0 <= s0
inserting runtime assert 0 <= -i0 + s0
inserting runtime assert i0 + i1 <= s0
inserting runtime assert i0 <= -i1 + s0
inserting runtime assert i0 + i1 + i2 <= s0
inserting runtime assert i0 + i1 <= -i2 + s0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
inserting runtime assert i0 + i1 + i2 + i3 <= s0
inserting runtime assert i0 + i1 + i2 <= -i3 + s0
```
to
```
inserting runtime assert i0 - s0 <= 0
inserting runtime assert i0 + i1 - s0 <= 0
inserting runtime assert i0 + i1 + i2 - s0 <= 0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
inserting runtime assert i0 + i1 + i2 + i3 - s0 <= 0
```

[ghstack-poisoned]

lezcano committed:
Update on "Canonicalize the runtime asserts"


This allows us to remove quite a few redundant runtime asserts. For
```
python test/dynamo/test_subclasses.py -k test_unbind
```

We go from
```
inserting runtime assert i0 <= s0
inserting runtime assert 0 <= -i0 + s0
inserting runtime assert i0 + i1 <= s0
inserting runtime assert i0 <= -i1 + s0
inserting runtime assert i0 + i1 + i2 <= s0
inserting runtime assert i0 + i1 <= -i2 + s0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
inserting runtime assert i0 + i1 + i2 + i3 <= s0
inserting runtime assert i0 + i1 + i2 <= -i3 + s0
```
to
```
inserting runtime assert i0 - s0 <= 0
inserting runtime assert i0 + i1 - s0 <= 0
inserting runtime assert i0 + i1 + i2 - s0 <= 0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
```

[ghstack-poisoned]

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -138,6 +138,54 @@ def is_concrete_int(a: Union[int, SymInt]):
+        return expr
+
+    if isinstance(expr, (sympy.And, sympy.Or, sympy.Not)):
+        expr = sympy.logic.boolalg.to_cnf(expr)
```
1) Can this logical expression simplify to False/True?

lezcano replied:
1. I wouldn't expect so. 2., I think it's fine with `simplify=False` as `simplify=True` looks like it does things that are too fancy and potentially expensive.

Chances are the original expression is already very close to being in CNF form, so this would just push `Neg`s inside and change a few ors for ands and and a few ands for ors.

lezcano replied:
I triggered a benchmark run and everything looks reasonably within noise. If this ever becomes an issue, we can implement our own simplified `to_cnf`.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -138,6 +138,54 @@ def is_concrete_int(a: Union[int, SymInt]):
+        lhs = expr.lhs - expr.rhs
+        t = type(expr)
+    rhs = 0
+    if isinstance(lhs, sympy.Add):
```
Why Add only for the constant shuffle?

lezcano replied:
Because we just want to solve for additive constants in expressions like to leave expressions like `2*x0 + x1 <= 9`. We could very well leave this on the left tbh. This is more aesthetics than anything.

lezcano replied:
Well, aesthetics and save a runtime addition / substraction in every of these guards.

lezcano committed:
Update on "Canonicalize runtime asserts"


This allows us to remove quite a few redundant runtime asserts, and potentially a number of guards as well.

On
```
python test/dynamo/test_subclasses.py -k test_unbind
```
we go from
```
inserting runtime assert i0 <= s0
inserting runtime assert 0 <= -i0 + s0
inserting runtime assert i0 + i1 <= s0
inserting runtime assert i0 <= -i1 + s0
inserting runtime assert i0 + i1 + i2 <= s0
inserting runtime assert i0 + i1 <= -i2 + s0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
inserting runtime assert i0 + i1 + i2 + i3 <= s0
inserting runtime assert i0 + i1 + i2 <= -i3 + s0
```
to
```
inserting runtime assert i0 - s0 <= 0
inserting runtime assert i0 + i1 - s0 <= 0
inserting runtime assert i0 + i1 + i2 - s0 <= 0
inserting runtime assert Eq(i0 + i1 + i2 + i3, s0)
```

cc ezyang

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
